scico/test/linop/test_radon_astra.py

```diff
@@ -95,4 +95,4 @@ def test_adjoint_grad(testobj):
 
 def test_adjoint(testobj):
     A = testobj.A
-    adjoint_test(A, rtol=get_tol())
+    adjoint_test(A, rtol=15 * get_tol())
```

Comment:
I'm now skeptical of the use of relative error in this test.

I believe the test boils down to relative error = |<y, Ax> - <Aty, x>| / max(<y, Ax>, <Aty, x>). But what if, e.g., A represents a 90 degree rotation? Then all these quantities are (nearly) zero and the resulting relative error is undefined/unstable.

Reply:
That is a good point. In the astra test context, it computes the dot product between a sinogram and the image. Given they are in two domains, their dot-product can be small. This doesn't make seem reasonable to me in a tomography context.

In the CT context we can divide by the norm of the image instead for the relative error. But not sure if that makes sense in the other cases the adjoint test is applied to.

Reply:
> I believe the test boils down to relative error = |<y, Ax> - <Aty, x>| / max(<y, Ax>, <Aty, x>). But what if, e.g., A represents a 90 degree rotation? Then all these quantities are (nearly) zero and the resulting relative error is undefined/unstable.

Sure about that? `<y, Ax>` is only small if `y` and `Ax` happen to be close to perpendicular (or if `A` is small but then also the numerator will be small).

> In the astra test context, it computes the dot product between a sinogram and the image.

Really? Seems like the dot product between _**a random**_ sinogram (`y`) and _**a random**_ projected image (`Ax`).

Reply:
By doing this way, all the tests pass.

Reply:
What is the actual numerical value of ``15 * get_tol()``? Performing a test like this with a relative error tolerance close to unity seems somewhat pointless.

Reply:
`15 * get_tol()` is 1.05. That makes it appear that the ASTRA backprojector is very inaccurate, but it's not the case as far as I know. Hence my skepticism of relative error.

Reply:
@tbalke good point, my 90 rotation example doesn't make sense because it is _not_ the case that y = Ax here.

Reply:
> > I believe the test boils down to relative error = |<y, Ax> - <Aty, x>| / max(<y, Ax>, <Aty, x>). But what if, e.g., A represents a 90 degree rotation? Then all these quantities are (nearly) zero and the resulting relative error is undefined/unstable.
> 
> Sure about that? `<y, Ax>` is only small if `y` and `Ax` happen to be close to perpendicular (or if `A` is small but then also the numerator will be small).
> 
> > In the astra test context, it computes the dot product between a sinogram and the image.
> 
> Really? Seems like the dot product between _**a random**_ sinogram (`y`) and _**a random**_ projected image (`Ax`).

Sorry, I was mistaken about that.
But I still think normalizing by the norm of the images instead of the dot product can be a better option for the relative tolerance. That way the denominator wouldn't get close to zero.

Reply:
Based on @Michael-T-McCann 's suggestion, I plotted  E[|<x,y>| / <x,x>]) vs number of dimensions where x,y are random normal vectors.

Seems to drop to zero pretty fast. So normalizing by the norm of the image rather than the dot product in the test can be a better normalization.

<img width="361" alt="Screen Shot 2021-11-19 at 2 57 29 PM" src="https://user-images.githubusercontent.com/16091159/142697118-05d19744-837b-4dd1-8317-b9d88628f568.png">

Reply:
Made changes to the normalization in commit e894d35. Let me know your thoughts on this. (haven't updated the docstring though).

Reply:
> ``15 * get_tol()`` is 1.05. That makes it appear that the ASTRA backprojector is very inaccurate, but it's not the case as far as I know. Hence my skepticism of relative error.

I agree that it doesn't look good for the ASTRA backprojector, but I'm not quite ready to just chalk this up as a failing of relative error.

Reply:
> Seems to drop to zero pretty fast. So normalizing by the norm of the image rather than the dot product in the test can be a better normalization.

> Made changes to the normalization in commit e894d35. Let me know your thoughts on this. (haven't updated the docstring though).

I'm not a fan of this approach because it doesn't normalize for changes in the norm of `A`, so a simple scaling of `A` would change our measure of adjoint accuracy. (But if one does take this approach, I think it would be better to just normalize the random `x` and `y` vectors after they're created.)

Reply:
I think it does normalize for changes in the norm of A. 

In commit e894d35 I normalized by max(||y||^2, ||Ax||^2). So scaling A would change both numerator and denominator for the relative error. But doing the max() may not be ideal. A better way might be normalizing by (||y|| ||Ax||)

Reply:
I didn't notice that in the commit; I thought you were just normalizing by the max of the norms of x and y. The max is really problematic because the scaling of the denominator with A depends on whether the norm of y is smaller than the norm of Ax. Normalizing by their product resolves that problem, but I'm not convinced that it's better than the original form, which seems to me to be more easily interpretable.

Reply:
The original form essentially normalizes by |y^T Ax|.
The problem with this is that for large random vectors y and Ax, the dot product can be fairly small since x,y are zero mean Gaussian vectors. This can make the relative error large. This issue wouldn't occur if x and y are constrained to be non-negative.

Reply:
Added new commit e2130b0

Reply:
Added the changes we discussed in the meeting.